documentation/specs/proposed/BuildCheck-Rules-Identification.md

```diff
@@ -0,0 +1,123 @@
+
+## Background and Context
+
+The MSBuild team is currently working on delivering the MSBuild analyzers (aka BuildCheck). The BuildCheck infrastructure has built-in analyzers and functionality to support custom ones. Hence, we need to make sure it will be possible to configure and differentiate built-in and custom analyzers.
+
+Note: Single analyzer can have multiple rules.
+
+### Problems to address:
+- The report should unambiguously point to the rule.
+- Execution and configuration issues and execution time reporting for analyzers.
+- Preventing clashes of identification within a single build and clashes of custom rules/analyzers with well-known rules/analyzers.
+- Possibility to configure the rule.
+
+## Proposal
+
+### Built-in analyzers
+Every built-in analyzer will have the friendly name: `BuildCheck.{FriendlyName}`.
+- Regular expression for the name: `^BuildCheck.[A-Z]{1,}[a-zA-Z0-9_]{0,}$`
+
+Each Rule that is shipped inbox will contain the RuleId as an identifier of the rule for this analyzer. 
+- The rule id format is as follows: `^BC[A-Za-z_.]{0,}[0-9]{1,}$`.
+
+#### Example of a built-in analyzer:
+- Name: `BuildCheck.SharedOutputPath`
+- RuleId: `BC0101` or `BC.AdditionalInfo0101` or `BC.Prefix.Test0123`
+
+### Custom analyzers
+Custom analyzer will have the friendly name: `{NameOfTheAnalyzer}` with defined format: 
+- `^[A-Z]{1,}[a-zA-Z_]{1,}$`
+- should not start with `BuildCheck.` this is built-in prefix for built-in.
+
+Each Custom Analyzer Rule will have the rule id format as follows:
+- `^[A-Z]{1}[A-Za-z]{0,}[0-9]{1,}$`. 
+- should not start from `BC` this is reserved prefix for built-in rules.
+
+#### Example of a custom analyzer:
+- Name: `SharedOutputPath`, `SharedOutputPath` 
+- RuleId: `SOMEPREFIX123`
+
+Any registered analyzers that don't follow the pattern (built-in and custom) will raise an exception and fail the build.
+
+The identification of the rule will consist of two components: the Friendlyname and the RuleId.
+
+#### Examples 
+- Built-in
+    - `BuildCheck.SharedOutputPath.BC0001`
+    - `BuildCheck.SharedOutputPath.BC0002`
+    - `BuildCheck.PropertyAssignedIncorrectly.BC0002`
+- Custom
+    - `SharedOutputPath.ID0001`
+    - `SharedOutputPath.ID0002`
+    - `PropertyAssignedIncorrectly.ID0002`
+
+#### Example of the output:
+```
+...
+Determining projects to restore...
+MSBUILD : error : BuildCheck.SharedOutputPath.BC0002: Projects FooBar-Copy.csproj and FooBar.csproj have onflicting output paths: C:\projects\msbuild\playground\buildcheck\bin\Debug\net8.0\.
+MSBUILD : error : BuildCheck.SharedOutputPath.BC0002: Projects FooBar-Copy.csproj and FooBar.csproj have onflicting output paths: C:\projects\msbuild\playground\buildcheck\obj\Debug\net8.0\.
+Restore:
+...
+```
+
+### Rules Identification clash prevention
+
+#### Custom VS Built-In
+The prevention of having the same analyzer/rule's name/id's between built-in and custom is guaranteed by preserved prefixes
+- Name Prefix: (BuildCheck|MSBuild|Microsoft)
```

Comment:
Even our team might want to possibly ship rules out of the band (orthogonal to official releases - to test-drive new checks).

I'm wondering what might be a good way out of a 3rd party pretending to be MSFT authored check, while allowing Microsoft teams (including us) to ship the checks via nugets as well... I suppose one thing we can rely on is the package name (as trustworthy feeds should already employ reserving the prefixes) - so we might allow the reserved Check prefix, only if it comes from package that has same prefix (we'd need to make sure to reserve the prefixes on nuget.org)

Reply:
Similar can apply to RuleId prefixes as well - basically "Only if you are build-in Check, or a Check comming from a package with some hardcoded recognized prefix (can be the 'Microsoft', 'BuildCheck', 'MSBuild' group) you can use any Check FriendlyName and RuleId prefixes you want, otherwise exclusions apply".

What do you think?

Reply:
`Microsoft*` is reserved on NuGet but I don't like the idea of coupling something in the rule engine (name validity) to NuGet in that way (I can't see a way of doing it other than "look in the path for a nuget package id pattern).

Reply:
Let's try to brainstorm options:
* Do not allow reserved prefixes in any custom analyzers (incl internaly produced)
* No checking of restricted prefixes
* Pigyback on nuget naming
* ?? (need ideas here :-))